<html>
<head>
<title>Annotation of Non-Coding RNA</title>
</head>

<body>
<h1>Non-coding RNA annotation</h1><h2>Data Sources</h2><p>If ncRNA gene annotations have been performed along with the protein coding genes predictions by the group or consortium in charge of a sequencing project, these annotations will be imported, either through GFF import or through INSDC import.</p><p>In the case there are no ncRNA annotations, we then run our own ncRNA predictions pipelines.</p><h2>Ensembl Genomes prediction pipelines</h2><p>For all <strong>ncRNA</strong> except <strong>tRNA</strong> and <strong>rRNA</strong> genes, models are predicted by aligning a genomic sequence against&nbsp;<a href="http://rfam.sanger.ac.uk/" rel="external">Rfam sequences</a> using <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi" rel="external">BLASTN</a>. The BLAST hits are then used to seed <a href="http://infernal.janelia.org" rel="external">Infernal</a> searches of the aligned regions with the corresponding Rfam covariance models. The purpose of this is to reduce the search space required, as to scan the entire genome with all the Rfam covariance models would be extremely CPU-intensive.</p><p>See Burge SW <em>et al.</em> (2013) <a href="http://europepmc.org/abstract/MED/23125362">Rfam 11.0: 10 years of RNA families</a> <em>Nucl. Acids Res.</em> <strong>41</strong> D226-32.</p><p><strong>tRNA</strong> is predicted by using <a href="http://selab.janelia.org/tRNAscan-SE">tRNAScan-SE</a> software. Version 1.23 of the program was used, configured for superregnum as appropriate.</p><p>See Lowe T.M. and Eddy S.R. (1997) <a href="http://europepmc.org/abstract/MED/9023104">tRNAScan-SE: a program for improved detection of transfer RNA genes in genomic sequence</a> <em>Nucl. Acids Res.</em> <strong>25</strong> 955-964</p><p><strong>rRNA</strong> is predicted by using <a href="http://www.cbs.dtu.dk/services/RNAmmer">RNAmmer</a> software. Version 1.2 of the program was used, configured for superregnum as appropriate.</p><p>See Lagesen K. <em>et al.</em> (2007) <a href="http://europepmc.org/abstract/MED/17452365">RNammer: consistent annotation of rRNA genes in genomic sequences</a> <em>Nucl. Acids Res.</em> <strong>35</strong> 3100-3108.</p><h2>Non-coding RNA biotype</h2><p>The following non-coding RNA gene types are annotated, along with pseudogenes.</p><dl><dt>tRNA</dt><dd>transfer RNA</dd><dt>rRNA</dt><dd>ribosomal RNA</dd><dt>scRNA</dt><dd>small cytoplasmic RNA</dd><dt>snRNA</dt><dd>small nuclear RNA</dd><dt>snoRNA</dt><dd>small nucleolar RNA</dd><dt>miRNA</dt><dd>microRNA precursors</dd><dt>tmRNA</dt><dd>transfer-messenger RNA</dd><dt>MRP_RNA</dt><dd>ribonuclease MRP (RNase MRP)</dd><dt>P_RNA</dt><dd>ribonuclease P (RNase P)</dd><dt>SRP_RNA</dt><dd>signal recognition particle (SRP) RNA</dd><dt>antisense</dt><dd>antisense RNA</dd><dt>ribozyme</dt><dd>ribozyme</dd><dt>telo_RNA</dt><dd>ribonucleoprotein reverse transcriptase (telomerase RNA)</dd><dt>v_RNA</dt><dd>vault RNA</dd><dt>class_I_RNA</dt><dt></dt><dd>class I RNA</dd><dt>class_II_RNA</dt><dt></dt><dd>class II RNA</dd><dt>misc_RNA</dt><dd>miscellaneous other RNA</dd></dl><p>Note that there are many more RFAM families but whenever they are classified as motifs (e.g. a SECIS element motif, <a href="http://rfam.sanger.ac.uk/family/RF00031">RF00031</a>), they are filtered out by our ncRNA gene prediction pipeline.</p>
</body>
</html>
